File: frontend/src/utils/cycleTraining.ts
import type { WorkoutPlan } from "@/types/workout";
import type { WeekPhase } from "@/utils/loadProgression";

export type CycleReadiness = "normal" | "caution" | "reduce" | "rest";

export const CYCLE_READINESS_OPTIONS: Array<{
  value: CycleReadiness;
  label: string;
  hint: string;
}> = [
  { value: "normal", label: "Всё хорошо", hint: "Оставить базовую нагрузку" },
  { value: "caution", label: "Без предельной нагрузки", hint: "Тяжёлая станет средней" },
  { value: "reduce", label: "Нужна лёгкая", hint: "Выполнить лёгкий вариант" },
  { value: "rest", label: "Нужно восстановление", hint: "Сначала предложим отложить" },
];

const FEMALE_SEX_VALUES = new Set(["f", "female", "woman", "ж", "жен", "женский", "женщина"]);
const UNSPECIFIED_SEX_VALUES = new Set(["", "unspecified", "not_specified", "не указан", "не указано"]);

export function cycleTrainingEnabledForProfile(
  goals: Record<string, unknown>,
  sex?: unknown,
): boolean {
  if (goals.cycle_training_enabled !== true) return false;
  const normalizedSex = String(sex || goals.sex || "").trim().toLowerCase();
  return FEMALE_SEX_VALUES.has(normalizedSex) || UNSPECIFIED_SEX_VALUES.has(normalizedSex);
}

export function isCycleReadiness(value: unknown): value is CycleReadiness {
  return ["normal", "caution", "reduce", "rest"].includes(String(value));
}

export function phaseFromPlan(plan: WorkoutPlan | null | undefined, fallback: WeekPhase): WeekPhase {
  const value = plan?.week_phase;
  return value === "light" || value === "medium" || value === "heavy" ? value : fallback;
}

File: backend/app/services/cycle_training.py
"""Optional, symptom-led menstrual-cycle training adjustments.

Calendar phase is deliberately not inferred: cycle length and individual response
vary too much for a fixed four-week prescription.  A user can instead record how
cycle-related symptoms affect today's readiness, and the planned load is capped
without changing the underlying program cursor.
"""

from __future__ import annotations

from typing import Any


CYCLE_READINESS_VALUES = {"normal", "caution", "reduce", "rest"}
PHASE_ORDER = {"light": 0, "medium": 1, "heavy": 2}


def normalize_cycle_readiness(value: object) -> str | None:
    normalized = str(value or "").strip().lower()
    return normalized if normalized in CYCLE_READINESS_VALUES else None


def cycle_training_enabled(
    goals: dict[str, Any] | None,
    anthropometry: dict[str, Any] | None = None,
) -> bool:
    """Allow the private option only for female or unspecified profiles."""

    if (goals or {}).get("cycle_training_enabled") is not True:
        return False
    raw_sex = (anthropometry or {}).get("sex") or (goals or {}).get("sex")
    sex = str(raw_sex or "").strip().lower().replace("ё", "е")
    is_male = sex in {"m", "male", "man", "м", "муж", "мужской"} or sex.startswith("муж")
    return not is_male


def adapt_week_phase(base_phase: str, readiness: object) -> dict[str, str] | None:
    """Return adaptation metadata, or ``None`` when the plan stays unchanged."""

    base = str(base_phase or "medium").strip().lower()
    if base not in PHASE_ORDER:
        base = "medium"
    value = normalize_cycle_readiness(readiness)
    if value in {None, "normal"}:
        return None

    if value == "caution":
        effective = "medium" if base == "heavy" else base
        label = "По самочувствию: без предельной нагрузки"
    elif value == "reduce":
        effective = "light"
        label = "По самочувствию: лёгкая нагрузка"
    else:
        effective = "light"
        label = "Сегодня приоритет — восстановление"

    if effective == base and value == "caution":
        return None
    return {
        "base_week_phase": base,
        "week_phase": effective,
        "load_adjustment": f"cycle_{value}",
        "load_adjustment_label": label,
    }
